pkg/types/dsse/v0.0.1/entry.go

```diff
@@ -0,0 +1,337 @@
+			}
+		}
+	default:
+		log.Logger.Infof("Cannot index payload of type: %s", payloadType)
+	}
+
+	return result, nil
+}
+
+func parseIntotoStatement(p []byte) (*in_toto.Statement, error) {
+	ps := in_toto.Statement{}
+	if err := json.Unmarshal(p, &ps); err != nil {
+		return nil, err
+	}
+
+	return &ps, nil
+}
+
+func (v *V001Entry) Unmarshal(pe models.ProposedEntry) error {
+	dsseModel, ok := pe.(*models.Dsse)
+	if !ok {
+		return errors.New("cannot unmarshal non DSSE v0.0.1 type")
+	}
+
+	if err := types.DecodeEntry(dsseModel.Spec, &v.DsseObj); err != nil {
+		return err
+	}
+
+	// field validation
+	if err := v.DsseObj.Validate(strfmt.Default); err != nil {
+		return err
+	}
+
+	if string(v.DsseObj.Payload) == "" {
+		return nil
+	}
+
+	env := &dsse.Envelope{
+		Payload:     string(v.DsseObj.Payload),
+		PayloadType: *v.DsseObj.PayloadType,
+	}
+
+	allPubKeyBytes := make([][]byte, 0)
+	for _, sig := range v.DsseObj.Signatures {
+		env.Signatures = append(env.Signatures, dsse.Signature{
+			KeyID: sig.Keyid,
+			Sig:   string(sig.Sig),
+		})
+
+		allPubKeyBytes = append(allPubKeyBytes, sig.PublicKey)
+	}
+
+	if _, err := verifyEnvelope(allPubKeyBytes, env); err != nil {
+		return err
+	}
+
+	return nil
+}
+
+func (v *V001Entry) Canonicalize(ctx context.Context) ([]byte, error) {
+	canonicalEntry := models.DsseV001Schema{
+		PayloadHash: v.DsseObj.PayloadHash,
+		Signatures:  v.DsseObj.Signatures,
+		PayloadType: v.DsseObj.PayloadType,
+	}
+
+	model := models.Dsse{}
+	model.APIVersion = swag.String(APIVERSION)
+	model.Spec = canonicalEntry
+	return json.Marshal(&model)
+}
+
+func (v *V001Entry) Attestation() []byte {
+	payload := v.DsseObj.Payload
+	if len(payload) > viper.GetInt("max_attestation_size") {
+		log.Logger.Infof("Skipping attestation storage, size %d is greater than max %d", len(payload), viper.GetInt("max_attestation_size"))
+		return nil
+	}
+
+	return payload
+}
+
+type verifier struct {
+	v signature.Verifier
+}
+
+func (v *verifier) KeyID() (string, error) {
+	return "", nil
+}
+
+func (v *verifier) Public() crypto.PublicKey {
+	// the dsse library uses this to generate a key ID if the KeyID function returns an empty string
+	// as well for the AcceptedKey return value.  Unfortunately since key ids can be arbitrary, we don't
+	// know how to generate a matching id for the key id on the envelope's signature...
+	// dsse verify will skip verifiers whose key id doesn't match the signature's key id, unless it fails
+	// to generate one from the public key... so we trick it by returning nil ¯\_(ツ)_/¯
+	return nil
+}
+
+func (v *verifier) Verify(data, sig []byte) error {
+	if v.v == nil {
+		return errors.New("nil verifier")
+	}
+	return v.v.VerifySignature(bytes.NewReader(sig), bytes.NewReader(data))
+}
+
+func (v V001Entry) CreateFromArtifactProperties(_ context.Context, props types.ArtifactProperties) (models.ProposedEntry, error) {
+	returnVal := models.Dsse{}
+	re := V001Entry{}
+
+	var err error
```

Comment:
Any reason to hoist this up here? It looks like it can be left inside the scope lower down.